### PR TITLE
cache buffer loads across multiple bufs

### PR DESCRIPTION
**Pull request by chaosagent**
Continuing #1303. Added test.

Cache loads by underlying rawbuffer, not by AST leaf. Helps in cases where upcasting results in the same load across two AST subtrees. Note that this does not CSE valid.max == 0 variables across differnt rawbufs -- this will be handled in a subsequent PR that folds these away.

Example:

Before (just running the added Linearizer test):

```
*** exec  0.00 GB    0.02 ms op: BinaryOps.ADD        out(float): (3,)                           in(2): [(3,)] 
UOps.DEFINE_GLOBAL  :                           []                               ('data0', dtypes.float)
UOps.DEFINE_GLOBAL  :                           []                               ('data1', dtypes.float)
   0 buffer<3, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   1 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   2 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=1, mask=None, contiguous=False, shape_strides=((3, 1),))]
   3
UOps.LOOP           :                           []                               ([], 'global')
UOps.LOOP           :                           []                               ([], 'local')
UOps.LOAD           : <val1_0>                  []                               MemOp(name='data1', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_1>                  []                               MemOp(name='data1', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_2>                  []                               MemOp(name='data1', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_0>                  []                               MemOp(name='data1', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_1>                  []                               MemOp(name='data1', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_2>                  []                               MemOp(name='data1', idx=<3>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ALU            : <alu0>                    [<val1_0>, <val2_0>]             BinaryOps.ADD
UOps.ALU            : <alu1>                    [<val1_1>, <val2_1>]             BinaryOps.ADD
UOps.ALU            : <alu2>                    [<val1_2>, <val2_2>]             BinaryOps.ADD
UOps.STORE          :                           [<alu0>]                         MemOp(name='data0', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu1>]                         MemOp(name='data0', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu2>]                         MemOp(name='data0', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ENDLOOP        :                           []                               ([], 'global+local')

```

After:

```
*** exec  0.00 GB    0.02 ms op: BinaryOps.ADD        out(float): (3,)                           in(2): [(3,)] 
UOps.DEFINE_GLOBAL  :                           []                               ('data0', dtypes.float)
UOps.DEFINE_GLOBAL  :                           []                               ('data1', dtypes.float)
   0 buffer<3, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   1 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   2 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=1, mask=None, contiguous=False, shape_strides=((3, 1),))]
   3
UOps.LOOP           :                           []                               ([], 'global')
UOps.LOOP           :                           []                               ([], 'local')
UOps.LOAD           : <val1_0>                  []                               MemOp(name='data1', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_1>                  []                               MemOp(name='data1', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_2>                  []                               MemOp(name='data1', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_3>                  []                               MemOp(name='data1', idx=<3>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ALU            : <alu0>                    [<val1_0>, <val1_1>]             BinaryOps.ADD
UOps.ALU            : <alu1>                    [<val1_1>, <val1_2>]             BinaryOps.ADD
UOps.ALU            : <alu2>                    [<val1_2>, <val2_3>]             BinaryOps.ADD
UOps.STORE          :                           [<alu0>]                         MemOp(name='data0', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu1>]                         MemOp(name='data0', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu2>]                         MemOp(name='data0', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ENDLOOP        :                           []                               ([], 'global+local')


```

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py            10               7
------------------------------------------------------------
lines added in the tinygrad folder: 3
```

**Comment by chaosagent** [comment]
An alternative way to implement this set of PRs exists: when upcasting, instead of placing a list of n tokens at the leaves of the tree, make n trees and use AST simplifiers to simplify them. I think this can get most of not all of the load cache / CSE / zero fold functionality. However, this changes the order that operations are done in.

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -224,7 +224,8 @@ def get_upcast_dim(self, i) -> List[int]:
     should_upcast = self.opts.supports_float4 and (self.bufs[i].dtype in [dtypes.float32, dtypes.float16] or isinstance(self.bufs[i].dtype, ImageDType))
     return [x for x in self.sts[i].unit_stride_axes() if should_upcast and x >= self.shape_len-self.upcasted and self.sts[i].shape[x] > 1]
 
-  def global_load(self, i:int, idxs:Sequence[VariableOrNum], const=None) -> List[Token]:
+  def global_load(self, i:int, idxs:Sequence[VariableOrNum], const=None, load_cache=None) -> List[Token]:
+    load_type = "val" if const is None else "acc"
```
Comment (geohot) [inline tinygrad/codegen/linearizer.py:228]: Why this line?

File: tinygrad/codegen/linearizer.py [hunk 2]
```diff
@@ -159,7 +159,7 @@ def __init__(self, ast:LazyOp, output_buffer:LazyBuffer, opts:LinearizerOptions)
 
   def get_buffer_name(self, i):
     if self.bufs[i].__class__ == LocalBuffer: return self.bufs[i].name
-    assert self.bufs[i].realized.__class__ is not RawConst  # constants shouldn't be loaded with memops
```
Comment (geohot) [inline tinygrad/codegen/linearizer.py:162]: Why should constants be loaded with memops now?

File: tinygrad/codegen/linearizer.py (hunk 1 above)
Comment (chaosagent) [reply tinygrad/codegen/linearizer.py:228]: it would write acc<num>_<num> for a RawConst

File: tinygrad/codegen/linearizer.py [hunk 3]
```diff
@@ -402,7 +403,8 @@ def ssa(name, ltype=dtypes.float) -> Token:
           self.uop(UOps.BARRIER, None, [], ())
 
         # load earlybufs
-        loaded_buffers.update({b:self.global_load(self.bufs.index(self.local_alias[i]) if i in self.local_alias else i, global_idxs+local_idxs+reduce_idxs+full_upcast_idxs) for i,b in enumerate(self.bufs) if b in self.earlybufs and i != 0})
+        earlybuf_cache: Dict[Any, Dict[str, Token]] = defaultdict(dict)
```
Comment (geohot) [inline tinygrad/codegen/linearizer.py:406]: Do you need separate caches?

Comment (chaosagent) [reply tinygrad/codegen/linearizer.py:406]: I ran into a case somewhere in the tests where a there was the same load in earlybufs and latebufs, but the earlybufs load was in the loop and out of scope of the late ast.

Comment (geohot) [reply tinygrad/codegen/linearizer.py:406]: kk, two cache is fine

File: tinygrad/codegen/linearizer.py (hunk 1 above)
Comment (geohot) [reply tinygrad/codegen/linearizer.py:228]: I don't see how it's different?

File: tinygrad/codegen/linearizer.py [hunk 4]
```diff
@@ -235,7 +236,7 @@ def global_load(self, i:int, idxs:Sequence[VariableOrNum], const=None) -> List[T
     if len(upcast_dim) == 1 and len(expanded_nodes[upcast_dim[0]]) in [4,2]:
       dim, amt = upcast_dim[0], len(expanded_nodes[upcast_dim[0]])
 
-    cache: Dict[str, Token] = {}
+    cache: Dict[str, Token] = load_cache[self.get_buffer_name(i)] if load_cache is not None and load_type == "val" else {}
```
Comment (geohot) [inline tinygrad/codegen/linearizer.py:239]: I'd do self.load_cache instead of passing it in. Like self.uop. And I don't think it should ever be None.

Comment (chaosagent) [reply tinygrad/codegen/linearizer.py:239]: Why would we want to cache accumulator definitions?

**Comment by chaosagent** [comment]
r4r again, maybe?

**Comment by geohot** [comment]
This works, merged!

**Pull request merged**